composer/models/gpt2/gpt2_hparams.py

```diff
@@ -3,63 +3,62 @@
     """`YAHP <https://docs.mosaicml.com/projects/yahp/en/stable/README.html>`_ interface for :class:`.GPT2Model`.
 
     Args:
-        pretrained_model_name (str): Pretrained model name to pull from Hugging Face Model Hub.
         model_config (Dict[str, JSON]): A dictionary providing a HuggingFace model configuration.
-        tokenizer_name (Optional[str]): The tokenizer used for this model,
-            necessary to assert required model inputs.
         use_pretrained (bool, optional): Whether to initialize the model with the pretrained weights. Default: ``False``.
+        tokenizer_name (Optional[str]): The tokenizer used for this model,
+            necessary to assert required model inputs. Default ``None``.
         gradient_checkpointing (bool, optional): Use gradient checkpointing. Default: ``False``.
     """
+    model_config: Optional[Dict[str,
+                                JSON]] = hp.optional(doc="A dictionary providing a HuggingFace model configuration.",
+                                                     default_factory=dict)
+    use_pretrained: Optional[bool] = hp.optional("Whether to initialize the model with the pretrained weights.",
+                                                 default=False)
+    tokenizer_name: Optional[str] = hp.optional(
+        "The tokenizer used for this model, necessary to assert required model inputs.", default=None)
+    gradient_checkpointing: Optional[bool] = hp.optional("Whether to enable gradient checkpointing.", default=False)
 
-    def initialize_object(self) -> "ComposerTransformer":
+    def initialize_object(self):
         try:
             import transformers
         except ImportError as e:
             raise MissingConditionalImportError(extra_deps_group="nlp", conda_package="transformers") from e
 
-        from composer.models.gpt2.model import GPT2Model
-        self.validate()
+        from composer.models.gpt2.model import create_gpt2
 
-        if self.model_config:
-            config = transformers.GPT2Config.from_dict(self.model_config)
-        elif self.pretrained_model_name is not None:
-            # TODO (Moin): verify that the config is an appropriate instance of GPT2!
-            config = transformers.GPT2Config.from_pretrained(self.pretrained_model_name)
-        else:
-            raise ValueError('One of pretrained_model_name or model_config needed.')
+        # user must specify one of either config or the pretrained model
+        if not self.use_pretrained and self.model_config == {}:
+            raise Exception('One of use_pretrained or model_config needed.')
+
+        if self.use_pretrained and self.model_config:
+            raise Exception('A model cannot load pretrained weights from configuration.')
 
         # setup the tokenizer in the hparams interface
-        if self.tokenizer_name is not None:
+        if self.tokenizer_name:
             tokenizer = transformers.GPT2Tokenizer.from_pretrained(self.tokenizer_name)
```

Comment:
can we move this logic to the factory functions? I'd like to keep interface parity between them and these hparams classes.

Reply:
same with the bert hparams. I think it's a better design to have all the logic for creating a model be in one place (factory function). and for the non-yaml and yaml args to be as closely aligned as possible.